Register Saved Tensors hooks

Varal7 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60975
* **#60663**

**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
At this point,  from the perspective of the user, `register_hooks` is still a no-op. 
The purpose of the PR is to create the class that will store the hooks. The logic of packing / unpacking will be added in the next PR.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

Varal7 committed:
Register Saved Tensors hooks

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -153,27 +153,31 @@ Variable SavedVariable::unpack(std::shared_ptr<Node> saved_for) const {
+  TORCH_CHECK(!hooks_,
+    "Calling register_hooks on a saved tensor whose hooks have already been set.");
+  hooks_ = std::move(hooks);
+  data_ = hooks_->call_unpack_hook(hooks_->call_pack_hook(data_));
```
I think you will need to use a new PyObject field here.
Also you want to make sure to take a refcount to the returned pyobject to make sure it is not freed.

Varal7 replied:
Oh yeah sorry, that comes in the next PR

Varal7 replied:
I should have left a comment saying so

Varal7 replied:
Sorry, maybe I should add a comment explaining that this line is temporary and will disappear in #60975

Varal7 committed:
Update on "Register Saved Tensors hooks"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466223).

Varal7 committed:
Update on "Register Saved Tensors hooks"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466223).

Varal7 committed:
Update on "Register Saved Tensors hooks"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466223).

Varal7 commented on torch/csrc/autograd/saved_variable_hooks.h:
```diff
@@ -0,0 +1,14 @@
+namespace torch { namespace autograd {
+
+struct TORCH_API SavedVariableHooks {
+  virtual PyObject* call_pack_hook(at::Tensor &tensor) = 0;
```
So the idea with this interface is to permit C++-only builds to not include any of the `PySavedVariableHooks` logic. 
It's fine because `SavedVariable::register_hooks` is only ever called from Python builds, but this way we don't need to include `PySavedVariableHooks.h` in `SavedVariable.h`

soulitzer replied:
Gotcha, maybe name it something like "PySavedVariableHooksInterface" to indicate that it's still python specific. Also you can add this explanation as a comment in that file.

Varal7 replied:
I would argue in favor of `SavedVariableHooks`, to honor the trend in the codebase
```c++
// python_hook.h
struct PyFunctionPreHook : public FunctionPreHook {
   // ...
}

// python_hook.h
struct PyFunctionPostHook : public FunctionPostHook {
  // ...
}

// python_engine.h
struct PythonEngine : public Engine {
  // ...
}

// python_function.h
struct PyNode : public Node {
 // ...
}

// python_anomaly_mode.h
struct PyAnomalyMetadata : public AnomalyMetadata {
  // ...
}
```

In particular, `AnomalyMetadata` is employed with the same strategy: it is used in the declaration of the virtual method `Engined::make_anomaly_metadata` whose only implementation is in `python_engine.cpp`. 
For some reason I haven't really dug into, AnomalyMetadata is also a concept that requires Python.

soulitzer replied:
But AnomalyMetadata does have a non-python implementation no?
https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/anomaly_mode.cpp#L43
Same for all the other examples, all of them can be used in a non-python context

Varal7 replied:
In the case of anomaly mode, the feature was first introduced as Python-only: https://github.com/pytorch/pytorch/pull/7677/files

soulitzer replied:
I think its okay to call it SavedVariableHooks if we make it Python-agnostic as the comment says
For example have `call_pack_hook` and `call_unpack_hook` take and return Tensors (maybe this was what you planned? I see that you did not import `python_headers.h` in this file.)
Otherwise, if we leave it in its current state, its probably better for the name to reflect the non-python-agnosticness.

Varal7 replied:
It's already technically python-agnostic thanks to [this](https://github.com/pytorch/pytorch/blob/master/c10/util/python_stub.h#L4), but I agree: I will rework this to make it easier to subclass if we ever want to implement hooks for C++ frontend.

soulitzer replied:
Oh interesting, didn't know about that stub! I guess there's no physical dependence here then - which I guess is "agnostic" some sense.

albanD replied:
Note from our latest discussion with Victor:
To make this future-proof for an eventual c++ implementation, we most likely dont to make the pack return nothing and the unpack take no input argument.
It is the SavedVariableHooks subclass responsibility to store any data internally. The python version will have a single py::object field for that while a c++ hook will be able to have any type as it sees fit.

Varal7 replied:
Yes, that's what ends up being done in #60975. 
I will probably reorganize the stack to make it easier to review.

Varal7 commented on torch/csrc/autograd/init.cpp:
```diff
@@ -269,8 +270,8 @@ PyObject* THPAutograd_initExtension(PyObject* _unused, PyObject *unused) {
       TORCH_CHECK(false, "Trying to create a SavedTensor object from Python is forbidden.");
     }))
     .def("register_hooks", [](torch::autograd::SavedVariable &s, py::function &pack_hook, py::function &unpack_hook) {
```
Type checking works as one would expect:

![image](https://user-images.githubusercontent.com/8019486/123886021-8fcfbd00-d91c-11eb-9fdb-fad294a69053.png)

I will add tests for that.

Varal7 committed:
Update on "Register Saved Tensors hooks"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466223).

Varal7 committed:
Update on "Register Saved Tensors hooks"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


Summary: This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. The current behavior of `register_hook` is to call `unpack(pack(data))`. The purpose is to check that packing/unpacking works correctly without handling logic about metadata and grads.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466223).

Varal7 commented:
@soulitzer I have investigated how pybind11 handles refcounting. 

`py::function` inherits from `py::object` as seen [here](https://github.com/pybind/pybind11/blob/6d4409466b6c7e6fa53e7b93cb05d78647c26313/include/pybind11/pytypes.h#L1458).
When we [call](https://github.com/pytorch/pytorch/pull/60663/files#diff-c58a67c85191c22c9b3bb439117d8053edfd9dea839fa010cf967d404c3c630dR276) `register_hooks`, a new `PySavedVariableHooks` is created which uses the copy-constructors of the two arguments `py::function pack_hook` and `py::function unpack_hook`. The copy-constructor is defined in the [parent class](https://github.com/pybind/pybind11/blob/6d4409466b6c7e6fa53e7b93cb05d78647c26313/include/pybind11/pytypes.h#L238)  and calls `inc_ref` which [increments the refcount](https://github.com/pybind/pybind11/blob/6d4409466b6c7e6fa53e7b93cb05d78647c26313/include/pybind11/pytypes.h#L192) of the underlying python function. When the `PySavedVariableHooks` is destructed, it calls [the destructor](https://github.com/pybind/pybind11/blob/6d4409466b6c7e6fa53e7b93cb05d78647c26313/include/pybind11/pytypes.h#L242) of the hook functions, which calls `dec_ref`.
So, in that sense, the functions are "kept alive" by the Saved Variable, but it uses a simpler mechanic than the `keep_alive` policy.

I added [a comment](https://github.com/pytorch/pytorch/pull/60663/files#diff-c58a67c85191c22c9b3bb439117d8053edfd9dea839fa010cf967d404c3c630dR273) to summarize this.

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
This PR changes the behavior of `register_hooks` from being a no-op to calling `unpack(pack(data))`, which is a costly no-op, and at this point, users still have no reason to call it. The behavior will change again in the next PR.
The purpose is to check that packing/unpacking works correctly without handling logic about metadata and grads.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466223).

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
This PR changes the behavior of `register_hooks` from being a no-op to calling `unpack(pack(data))`, which is a costly no-op, and at this point, users still have no reason to call it. The behavior will change again in the next PR.
The purpose is to check that packing/unpacking works correctly without handling logic about metadata and grads.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466223).

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
This PR changes the behavior of `register_hooks` from being a no-op to calling `unpack(pack(data))`, which is a costly no-op, and at this point, users still have no reason to call it. The behavior will change again in the next PR.
The purpose is to check that packing/unpacking works correctly without handling logic about metadata and grads.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
This PR changes the behavior of `register_hooks` from being a no-op to calling `unpack(pack(data))`, which is a costly no-op, and at this point, users still have no reason to call it. The behavior will change again in the next PR.
The purpose is to check that packing/unpacking works correctly without handling logic about metadata and grads.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466223).

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
This PR changes the behavior of `register_hooks` from being a no-op to calling `unpack(pack(data))`, which is a costly no-op, and at this point, users still have no reason to call it. The behavior will change again in the next PR.
The purpose is to check that packing/unpacking works correctly without handling logic about metadata and grads.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
This PR changes the behavior of `register_hooks` from being a no-op to calling `unpack(pack(data))`, which is a costly no-op, and at this point, users still have no reason to call it. The behavior will change again in the next PR.
The purpose is to check that packing/unpacking works correctly without handling logic about metadata and grads.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
This PR changes the behavior of `register_hooks` from being a no-op to calling `unpack(pack(data))`, which is a costly no-op, and at this point, users still have no reason to call it. The behavior will change again in the next PR.
The purpose is to check that packing/unpacking works correctly without handling logic about metadata and grads.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
This PR changes the behavior of `register_hooks` from being a no-op to calling `unpack(pack(data))`, which is a costly no-op, and at this point, users still have no reason to call it. The behavior will change again in the next PR.
The purpose is to check that packing/unpacking works correctly without handling logic about metadata and grads.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_saved_variable_hooks.h:
```diff
@@ -0,0 +1,31 @@
+    PySavedVariableHooks(py::function &pack_hook, py::function &unpack_hook) : pack_hook_(pack_hook), unpack_hook_(unpack_hook){}
+
+    PyObject* call_pack_hook(at::Tensor &tensor) override {
+      return pack_hook_(py::reinterpret_steal<py::object>(THPVariable_Wrap(tensor))).release().ptr();
```
note that this one liner makes it tricky to know when refcount bumps happen for the python objects.
You might want to split this into 3 lines to make it clear that you first get a THPVariable for which you own a reference, then you pass that to the hook function which returns another object with a reference. You then release the new object on return and the THPVariable refcount is decrefed on exit by the py::object destruction.

Varal7 replied:
Good point. Splitting in 3 lines made me realize that the `call_unpack_hook` is actually incorrect in its current form

albanD commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -167,14 +167,22 @@ void SavedVariable::register_hooks() {
     }
   }
+  TORCH_CHECK(!hooks_,
+    "Calling register_hooks on a saved tensor whose hooks have already been set.");
```
nit: mention that only one hook at a time can be used here.

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
This PR changes the behavior of `register_hooks` from being a no-op to calling `unpack(pack(data))`, which is a costly no-op, and at this point, users still have no reason to call it. The behavior will change again in the next PR.
The purpose is to check that packing/unpacking works correctly without handling logic about metadata and grads.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
At this point,  from the perspective of the user, `register_hooks` is still a no-op. 
The purpose of the PR is to create the class that will store the hooks. The logic of packing / unpacking will be added in the next PR.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
At this point,  from the perspective of the user, `register_hooks` is still a no-op. 
The purpose of the PR is to create the class that will store the hooks. The logic of packing / unpacking will be added in the next PR.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
At this point,  from the perspective of the user, `register_hooks` is still a no-op. 
The purpose of the PR is to create the class that will store the hooks. The logic of packing / unpacking will be added in the next PR.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
At this point,  from the perspective of the user, `register_hooks` is still a no-op. 
The purpose of the PR is to create the class that will store the hooks. The logic of packing / unpacking will be added in the next PR.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
At this point,  from the perspective of the user, `register_hooks` is still a no-op. 
The purpose of the PR is to create the class that will store the hooks. The logic of packing / unpacking will be added in the next PR.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
At this point,  from the perspective of the user, `register_hooks` is still a no-op. 
The purpose of the PR is to create the class that will store the hooks. The logic of packing / unpacking will be added in the next PR.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
At this point,  from the perspective of the user, `register_hooks` is still a no-op. 
The purpose of the PR is to create the class that will store the hooks. The logic of packing / unpacking will be added in the next PR.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
At this point,  from the perspective of the user, `register_hooks` is still a no-op. 
The purpose of the PR is to create the class that will store the hooks. The logic of packing / unpacking will be added in the next PR.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
At this point,  from the perspective of the user, `register_hooks` is still a no-op. 
The purpose of the PR is to create the class that will store the hooks. The logic of packing / unpacking will be added in the next PR.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_saved_variable_hooks.cpp:
```diff
@@ -0,0 +1,27 @@
+  }
+
+  PySavedVariableHooks::~PySavedVariableHooks() {
+      // If python is already dead, leak the wrapped python objects
```
nit identation is not aligned

albanD reviewed (commented):
Looks good on my end.
I'll let @soulitzer get a final look

soulitzer commented on torch/csrc/autograd/python_saved_variable_hooks.cpp:
```diff
@@ -0,0 +1,27 @@
+  PySavedVariableHooks::~PySavedVariableHooks() {
+      // If python is already dead, leak the wrapped python objects
+    if (Py_IsInitialized()) {
+      pybind11::gil_scoped_acquire gil;
```
nit: py::gil_scoped_acquire since you already aliased the namespace?

soulitzer commented on torch/csrc/autograd/python_saved_variable_hooks.h:
```diff
@@ -0,0 +1,25 @@
+  ~PySavedVariableHooks() override;
+
+private:
+  PyObject* pack_hook_;
```
Is the reason we can't store `py::function` directly that we need to leak it on destruction sometimes? It's interesting that pybind doesn't handle this case.

albanD replied:
No, we could actually properly release the py::functions in the destructor the same way.
The problem is that pybind is imported with visibility hidden (from the c++ definition) and so that means that we cannot use them as attributes on a class that is visible (due to the TORCH_API). You basically say that the class should be exposed but the attribute's type hidden which doesn't work.
Moving to PyObject (that is always exposed) is a simple fix for that. We might be able to make py::function exposed as well, but I don't know how from the top of my head.

soulitzer reviewed (approved):
Looks good to me also!

soulitzer commented:
Also what's up with the TORCH_API/THP_CLASS macros, I noticed that you were tweaking that in the past couple iterations. What was the idea behind those changes?

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
At this point,  from the perspective of the user, `register_hooks` is still a no-op. 
The purpose of the PR is to create the class that will store the hooks. The logic of packing / unpacking will be added in the next PR.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466223).

Varal7 commented:
Thanks for the review @soulitzer and thank you @albanD for helping me debug this PR.

So `TORCH_API` is the macro we use to set the visibility of classes / methods for windows builds. However, each library uses its own `*something*_API`. For instance,`TORCH_API` is only for `libtorch_core` but we need to have our `python_saved_variable_hooks.h` in `libtorch_python`. There is no such thing as `TORCH_PYTHON_API` but `THP_API` is used in things like `python_variable.h`. Sadly, this failed too for the reason that Alban stated above: `py::function` is not exposed by pybind, so more visibility problems.
The current form bypasses all problems. 

For posterity: if you have linking problems with symbols not found:
1./Try adding a `TORCH_API` 
2/ Move your method definitions outside the header and into the body (cpp file)
3/ Make sure `build_variables.bzl` and `BUILD.bazel` are correct
4/ Are you using the incorrect `*library*_API` macro, should you move your file to a different library?

If you want to debug windows builds and want to ssh 
1/ If you have a visibility / linking problem, see above.
2/ You can tag your PR with the ci/master label which triggers CircleCI tests and allows for "rerun with SSH"

Varal7 committed:
Update on "Register Saved Tensors hooks"


**Summary**: 
This PR is part of a stack aimed to address #58512.
Saves the hooks as a field of `SavedVariable` with type `std::unique_ptr<SavedVariableHooks>`. 
At this point,  from the perspective of the user, `register_hooks` is still a no-op. 
The purpose of the PR is to create the class that will store the hooks. The logic of packing / unpacking will be added in the next PR.


Differential Revision: [D29466223](https://our.internmc.facebook.com/intern/diff/D29466223)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466223).
